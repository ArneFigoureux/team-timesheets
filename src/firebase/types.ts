export enum LoginProvider {
    Email = 0,
    Facebook = 1,
    Google = 2,
    Guest = 3,
}
