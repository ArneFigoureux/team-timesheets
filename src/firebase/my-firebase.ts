import "firebase/firestore";
import * as firebase from "firebase/app";
import { config } from "../config";

const app = firebase.initializeApp({
    apiKey: config.apiKey,
    authDomain: config.authDomain,
    projectId: config.projectId,
    storageBucket: config.storageBucket
});

export const firestore = app.firestore();
export const storage = app.storage();
export const auth = app.auth();
export const functions = app.functions();