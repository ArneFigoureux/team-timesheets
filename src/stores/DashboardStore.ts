import { observable, reaction, computed, IObservableValue, action } from "mobx";
import { ICollection, Collection } from "../Firestorable/Collection";

import { IRootStore } from "./RootStore";
import { IRegistration, IRegistrationData, IGroupedRegistrations } from "./TimesheetsStore";
import * as deserializer from '../serialization/deserializer';
import * as serializer from '../serialization/serializer';

export interface IDashboardStore {
    readonly setProjectFilter: (projectId: string | undefined) => void;
    readonly setUserFilter: (userId: string | undefined) => void;
    readonly setTaskFilter: (taskId: string | undefined) => void;

    readonly registrationsGroupedByUser: IGroupedRegistrations<string>[];
    readonly registrationsGroupedByTask: IGroupedRegistrations<string>[];
    readonly registrationsGroupedByProject: IGroupedRegistrations<string>[];

    readonly projectFilterValue?: string;
    readonly taskFilterValue?: string;
    readonly userFilterValue?: string;
}

export class DashboardStore implements IDashboardStore {
    private readonly registrationsField: ICollection<IRegistration>;

    @observable private startDate: IObservableValue<Date | undefined> = observable.box();
    @observable private endDate: IObservableValue<Date | undefined> = observable.box();
    @observable.ref private userFilterValueField: string | undefined = undefined;
    @observable.ref private taskFilterValueField: string | undefined = undefined;
    @observable.ref private projectFilterValueField: string | undefined = undefined;

    constructor(rootStore: IRootStore) {
        this.registrationsField = observable(new Collection<IRegistration, IRegistrationData>(() => rootStore.getCollection("registrations"),
            {
                realtime: true,
                deserialize: deserializer.convertRegistration,
                serialize: serializer.convertRegistration
            }));

        const updateRegistrationQuery = () => {
            this.registrationsField.query = ref =>
                this.userIdFilter(
                    this.projectIdFilter(
                        this.endDateFilter(
                            this.startDateFilter(
                                ref.where("deleted", "==", false)
                            )
                        )
                    )
                );
        };

        // update the query of the registration collection each time...
        // -- a filter changes (userId, startDate, endDate)
        // TODO: add a Filter button in UI to group all filter changes into one new search query
        reaction(() => this.startDate, updateRegistrationQuery);
        reaction(() => this.endDate, updateRegistrationQuery);
        reaction(() => this.userFilterValue, updateRegistrationQuery);
        reaction(() => this.projectFilterValue, updateRegistrationQuery);

        this.registrationsField.getDocs();
    }

    @computed
    private get startDateFilter() {
        const startDate = this.startDate.get()
        return (query: firebase.firestore.Query) => startDate ? query.where("date", ">=", startDate) : query;
    }

    @computed
    private get endDateFilter() {
        const endDate = this.endDate.get()
        return (query: firebase.firestore.Query) => endDate ? query.where("date", "<=", endDate) : query;
    }

    @computed
    private get userIdFilter() {
        const userId = this.userFilterValue;
        return (query: firebase.firestore.Query) => userId ? query.where("userId", "==", userId) : query;
    }

    @computed
    private get projectIdFilter() {
        return (query: firebase.firestore.Query) => this.projectFilterValue ? query.where("project", "==", this.projectFilterValue) : query;
    }

    @computed
    get registrationsGroupedByUser() {
        // no need to group when userId is included in the filter
        if (this.registrations.length === 0) return [];

        const registrationsPerUser = this.createRegistrationsGroup(reg => reg.userId);

        return registrationsPerUser;
    }

    @computed
    get registrationsGroupedByProject() {
        if (this.registrations.length === 0) return [];

        return this.createRegistrationsGroup(reg => reg.project || "")
            .sort((a, b) => b.totalTime - a.totalTime);;
    }

    @computed
    get registrationsGroupedByTask() {
        if (this.registrations.length === 0) return [];

        return this.createRegistrationsGroup(reg => reg.task || "")
            .sort((a, b) => b.totalTime - a.totalTime);
    }

    public get taskFilterValue() {
        return this.taskFilterValueField;
    }

    public get userFilterValue() {
        return this.userFilterValueField;
    }

    public get projectFilterValue() {
        return this.projectFilterValueField;
    }

    @computed
    private get registrations() {
        return Array.from(this.registrationsField.docs.values())
            .filter(doc => doc.data!.isPersisted) // don't display drafts (TODO: move to collection.ts?)
    }

    @action
    public setProjectFilter(projectId: string | undefined) {
        this.projectFilterValueField = projectId;
    }

    @action
    public setTaskFilter(taskId: string | undefined) {
        this.taskFilterValueField = taskId;
    }

    @action
    public setUserFilter(userId: string | undefined) {
        this.userFilterValueField = userId;
    }

    private createRegistrationsGroup = <T>(keySelector: (reg: IRegistration) => T): IGroupedRegistrations<T>[] => {
        const groupedMap = new Map<T, IGroupedRegistrations<T>>();

        this.registrations
            .forEach(c => {
                const key = keySelector(c.data!);
                const group = groupedMap.get(key);
                if (group) {
                    group.totalTime += c.data!.time || 0;
                    group.registrations.push(c);
                }
                else {
                    groupedMap.set(key, {
                        groupKey: keySelector(c.data!),
                        registrations: [c],
                        totalTime: c.data!.time || 0
                    });
                }
            });

        return Array.from(groupedMap.values());
    }
}