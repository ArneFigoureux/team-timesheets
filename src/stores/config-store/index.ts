export * from "./config-store";
export * from "./use-config-store";