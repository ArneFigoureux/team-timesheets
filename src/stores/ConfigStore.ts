import { observable, computed } from 'mobx';
import { Collection, ICollection } from "firestorable";
import { IRootStore } from './RootStore';
import { IProject, ITask, IClient, ITeam } from '../../common/dist';
import { firestore } from '../firebase/myFirebase';

export interface IConfigStore {
    projects: ICollection<IProject>;
    tasks: ICollection<ITask>;
    clientsCollection: ICollection<IClient>;
    teamsCollection: ICollection<IClient>;
    clients: (IClient & {id: string})[];
    teams: (ITeam & {id: string, isSelected: boolean})[];
    taskId?: string;
    projectId?: string;
    clientId?: string;
    teamId?: string;
}

export class ConfigStore implements IConfigStore {
    //private readonly _rootStore: IRootStore;

    readonly projects: ICollection<IProject>;
    readonly tasks: ICollection<IProject>;
    readonly clientsCollection: ICollection<IClient>;
    readonly teamsCollection: ICollection<ITeam>;

    @observable.ref taskId?: string;
    @observable.ref projectId?: string;
    @observable.ref clientId?: string;
    @observable.ref teamId?: string;

    constructor(_rootStore: IRootStore, getCollection: (name: string) => firebase.firestore.CollectionReference) {
        // this._rootStore = rootStore;
        this.projects = observable(new Collection<IProject>(firestore, getCollection.bind(this, "projects"), {
            realtime: true,
            query: ref => ref.orderBy("name")
        }));

        this.tasks = observable(new Collection<ITask>(firestore, getCollection.bind(this, "tasks"), {
            realtime: true,
            query: ref => ref.orderBy("name")
        }));

        this.clientsCollection = observable(new Collection<ITask>(firestore, getCollection.bind(this, "clients"), {
            realtime: true,
            query: ref => ref.orderBy("name")
        }));

        this.teamsCollection = observable(new Collection<ITeam>(firestore, getCollection.bind(this, "teams"), {
            realtime: true,
            query: ref => ref.orderBy("name")
        }));
    }

    @computed
    public get clients() {
        return Array.from(this.clientsCollection.docs.values())
            .map(doc => ({ ...doc.data!, id: doc.id }));
    }

    @computed
    public get teams() {
        return Array.from(this.teamsCollection.docs.values())
            .map(doc => ({ ...doc.data!, id: doc.id, isSelected: doc.id === this.teamId }));
    }
}