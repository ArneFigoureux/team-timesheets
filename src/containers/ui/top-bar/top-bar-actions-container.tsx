import React from "react";
import { observer } from "mobx-react-lite";

import { TopBarActions as PureTopBarActions } from "./top-bar-actions";
import { useViewStore } from "../../../stores/view-store";

export const TopBarActions = observer(() => {
    const view = useViewStore();

    const contextual = !!Array.from(view.selection.keys()).length;

    const actions = view.actions.filter(a => !!a.contextual === contextual);

    return <PureTopBarActions actions={actions} />;
});