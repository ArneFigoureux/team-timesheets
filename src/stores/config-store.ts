import { observable, computed } from 'mobx';
import { Collection, ICollection, RealtimeMode, FetchMode } from "firestorable";
import { IRootStore } from './root-store';
import { firestore } from '../firebase/my-firebase';
import { IProject, ITask, IClient, IClientData, ITeam, ITeamData, ITaskData, IConfig, ConfigValue } from '../../common/dist';

import * as serializer from '../../common/serialization/serializer';
import * as deserializer from '../../common/serialization/deserializer';
import { LoginProvider } from '../firebase/types';

export type Configs = {
    // authClientId: string;
    loginProviders: LoginProvider[];
}

export interface IConfigStore extends ConfigStore {};

export class ConfigStore implements IConfigStore {
    //private readonly _rootStore: IRootStore;
    readonly tasks: ICollection<IProject, ITaskData>;
    readonly clientsCollection: ICollection<IClient>;
    readonly teamsCollection: ICollection<ITeam, ITeamData>;
    readonly configsCollection: Collection<IConfig>;

    @observable.ref taskId?: string;
    @observable.ref clientId?: string;
    @observable.ref teamId?: string;

    constructor(_rootStore: IRootStore) {
        // this._rootStore = rootStore;
        const deps = { logger: console.log };

        this.tasks = new Collection<ITask, ITaskData>(
            firestore,
            "tasks",
            {
                realtimeMode: RealtimeMode.on,
                fetchMode: FetchMode.once,
                query: ref => ref.orderBy("name_insensitive"),
                serialize: serializer.convertNameWithIcon,
                deserialize: deserializer.convertNameWithIcon,
            },
            deps,
        );

        this.clientsCollection =
            new Collection<IClient, IClientData>(
                firestore,
                "clients",
                {
                    realtimeMode: RealtimeMode.on,
                    fetchMode: FetchMode.once,
                    query: ref => ref.orderBy("name_insensitive"),
                    serialize: serializer.convertNameWithIcon,
                    deserialize: deserializer.convertNameWithIcon,
                },
                deps,
            );

        this.teamsCollection = new Collection<ITeam, ITeamData>(
            firestore,
            "teams",
            {
                realtimeMode: RealtimeMode.on,
                fetchMode: FetchMode.once,
                query: ref => ref.orderBy("name_insensitive"),
                serialize: serializer.convertNameWithIcon,
                deserialize: deserializer.convertNameWithIcon,
            },
            deps,
        );

        this.configsCollection = new Collection<IConfig>(
            firestore,
            "configs",
            {
                fetchMode: FetchMode.once,
            }
        )
    }

    @computed
    public get clients() {
        return Array.from(this.clientsCollection.docs.values())
            .map(doc => ({ ...doc.data!, id: doc.id }));
    }

    @computed
    public get teams() {
        return Array.from(this.teamsCollection.docs.values())
            .map(doc => ({ ...doc.data!, id: doc.id, isSelected: doc.id === this.teamId }));
    }

    // To investigate: does getConfigValue needs mobx @computed attribute?
    public getConfigValue<T>(key: string): T;
    public getConfigValue<T>(key: string, isRequired: false): T | undefined;
    public getConfigValue<T extends ConfigValue>(key: string, isRequired = true): T | undefined {
        const doc = this.configsCollection.docs.find(c => c.data!.key === key);

        if (!doc) {
            if (isRequired) throw new Error(`Required config '${key}' is missing.`);
            else return undefined;
        }

        return doc.data!.value as unknown as T;
    }
}
