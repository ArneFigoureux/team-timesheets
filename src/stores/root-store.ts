import { IRegistrationsStore, RegistrationStore } from "./registration-store";
import { IConfigStore, ConfigStore } from "./config-store";
import { IUserStore, UserStore } from "./user-store";
import { IViewStore, ViewStore } from "./view-store";
import { RouterStore } from "mobx-router";
import { IReportStore, ReportStore } from "./report-store";
import { DashboardStore, IDashboardStore } from "./dashboard-store";
import { firestore } from "../firebase/my-firebase";
import { IProjectStore, ProjectStore } from "./project-store";

export interface IRootStore {

    readonly user: IUserStore;
    readonly view: IViewStore;
    readonly router: RouterStore;
    readonly timesheets: IRegistrationsStore;
    readonly reports: IReportStore;
    readonly config: IConfigStore;
    readonly projects: IProjectStore;
    readonly getCollection: (name: string) => firebase.firestore.CollectionReference;
}

export class Store implements IRootStore {
    public readonly timesheets: IRegistrationsStore;
    public readonly view: IViewStore;
    public readonly user: IUserStore;
    public readonly config: IConfigStore;
    public readonly router = new RouterStore();
    public readonly reports: IReportStore;
    public readonly dashboard: IDashboardStore;
    public readonly projects: IProjectStore;

    public readonly getCollection: (name: string) => firebase.firestore.CollectionReference;

    constructor(getCollection: (name: string) => firebase.firestore.CollectionReference) {
        this.getCollection = getCollection;

        this.user = new UserStore(this);
        this.view = new ViewStore(this);
        this.config = new ConfigStore(this, getCollection);
        this.timesheets = new RegistrationStore(this);
        this.reports = new ReportStore(this);
        this.dashboard = new DashboardStore(this);
        this.projects = new ProjectStore(this);

    }
};

const store = (window as any)["store"] = new Store(name => firestore.collection(name));

export default store;

