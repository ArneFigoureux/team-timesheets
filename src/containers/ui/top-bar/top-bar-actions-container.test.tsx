import React from "react";
import { initTestFirestore, deleteFirebaseAppsAsync } from "../../../__tests__/utils/firebase";
import { Store } from "../../../stores/root-store";
import { render, fireEvent } from "@testing-library/react";
import { TopBarActions } from "./top-bar-actions-container";
import { ObservableMap } from "mobx";

const {
    firestore,
    clearFirestoreDataAsync,
} = initTestFirestore("top-bar-actions-test", []);

const store = new Store({
    firestore,
});

jest.mock("../../../contexts/store-context", () => ({
    useStore: () => store,
}));

beforeAll(clearFirestoreDataAsync);

afterAll(() => {
    store.dispose();
    return Promise.all([
        deleteFirebaseAppsAsync(),
    ])
});


describe("TopBarActionsContainer", () => {
    it("should render when there are no actions", () => {
        const { asFragment } = render(<TopBarActions />);

        expect(asFragment()).toMatchSnapshot();
    });

    it("should show actions", () => {
        store.view.setActions([
            {
                action: jest.fn(),
                icon: {
                    content: "favorite",
                    label: "Favorite",
                },
            },
            {
                action: jest.fn(),
                icon: {
                    content: "house",
                    label: "House",
                },
                isActive: false,
            },
            {
                action: jest.fn(),
                icon: {
                    content: "save",
                    label: "Save",
                },
                isActive: () => true,
            },
            {
                action: jest.fn(),
                icon: {
                    content: "people",
                    label: "People",
                },
                iconActive: {
                    content: "people_border",
                    label: "People active",
                },
                isActive: true,
            },
            {
                action: jest.fn(),
                icon: {
                    content: "close",
                    label: "Close",
                },
                selection: new ObservableMap(),
            },
        ]);

        const { getByText } = render(<TopBarActions />);

        expect(getByText("favorite")).toBeDefined();
        expect(getByText("people_border")).toBeDefined();
        store.view.setActions([]);
    });

    it("should not show contextual actions when there is no selection", () => {
        store.view.setActions([
            {
                action: jest.fn(),
                icon: {
                    content: "favorite",
                    label: "Favorite",
                },
                contextual: true,
            },
            {
                action: jest.fn(),
                icon: {
                    content: "people",
                    label: "People",
                },
                iconActive: {
                    content: "people_border",
                    label: "People active",
                },
                isActive: true,
            },
        ]);

        const { queryByText, getByText, asFragment } = render(<TopBarActions />);

        expect(asFragment()).toMatchSnapshot();
        expect(queryByText("favorite")).toBeNull();
        expect(getByText("people"));

        store.view.setActions([]);
    });

    it("should only show contextual actions when there is a selection", () => {
        store.view.setActions([
            {
                action: jest.fn(),
                icon: {
                    content: "favorite",
                    label: "Favorite",
                },
                contextual: true,
            },
            {
                action: jest.fn(),
                icon: {
                    content: "people",
                    label: "People",
                },
                iconActive: {
                    content: "people_border",
                    label: "People active",
                },
                isActive: true,
            },
        ]);

        store.view.selection.set("foo", true);

        const { queryByText, getByText, asFragment } = render(<TopBarActions />);

        expect(asFragment()).toMatchSnapshot();
        expect(getByText("favorite"));
        expect(queryByText("people")).toBeNull();

        store.view.setActions([]);
        store.view.selection.clear();
    });

    it("should call the action when clicked", () => {

        const action = jest.fn();
        store.view.setActions([
            {
                action,
                icon: {
                    content: "favorite",
                    label: "Favorite",
                },
            },
        ]);

        const { getByText } = render(<TopBarActions />);

        fireEvent.click(getByText("favorite"));

        expect(action).toBeCalled();

        store.view.setActions([]);
        store.view.selection.clear();
    });
});