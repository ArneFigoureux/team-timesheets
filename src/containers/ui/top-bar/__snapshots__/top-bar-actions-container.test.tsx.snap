// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopBarActionsContainer should not show contextual actions when there is no selection 1`] = `
<DocumentFragment>
  <button
    aria-pressed="true"
    class="mdc-ripple-upgraded--unbounded mdc-top-app-bar__action-item mdc-icon-button--on mdc-icon-button mdc-icon-button--on"
    data-mdc-ripple-is-unbounded="true"
    role="button"
    tabindex="0"
  >
    <i
      class="rmwc-icon rmwc-icon--ligature material-icons mdc-icon-button__icon"
    >
      people
    </i>
    <i
      class="rmwc-icon rmwc-icon--ligature material-icons mdc-icon-button__icon mdc-icon-button__icon--on"
    >
      people_border
    </i>
  </button>
</DocumentFragment>
`;

exports[`TopBarActionsContainer should only show contextual actions when there is a selection 1`] = `
<DocumentFragment>
  <button
    class="rmwc-icon rmwc-icon--ligature material-icons mdc-ripple-upgraded--unbounded mdc-top-app-bar__action-item mdc-icon-button"
    data-mdc-ripple-is-unbounded="true"
    role="button"
    tabindex="0"
  >
    favorite
  </button>
</DocumentFragment>
`;

exports[`TopBarActionsContainer should render when there are no actions 1`] = `<DocumentFragment />`;
